CIP-array-builtin/README.md

```diff
@@ -0,0 +1,296 @@
+    - It also allows users to omit bounds checks when they know a priori that the
+    index is in bounds.
+
+After considering these three options, we concluded that the addition of
+the `lengthOfArray` builtin function is both necessary and sufficient for introducing a
+well-defined array type into the builtin language.
+
+### Constructing arrays
+
+The last required functionality for having a practical interface is the ability to
+construct arrays.
+
+Due to our decision of providing immutable arrays, it is difficult (or very expensive) to build up arrays incrementally.
+A naive approach would require repeated copying and potentially the usage of quadratic
+space and time.
+
+A more appropriate approach would be to construct the array in bulk,
+however that would require an intermediate representation of a collection of elements. Fortunately this already exists in the builtin language in the form of builtin lists.
+We can then naturally introduce a function which transforms lists into arrays: `listToArray`.
+
+Since the only current method of constructing arrays is via `listToArray`, one may ask
+whether the `lengthOfArray` builtin is at all necessary.
+We believe that, indeed, `lengthOfArray` is redundant in the current specification. The reason is
+that traversing the list to construct the array is enough to also compute its length. 
+However, we have decided that `lengthOfArray` should be kept as future work will depend upon its
+existence. Some components of Plutus Core could be natively represented as arrays,
+and there would be no other way to access their length in constant time other than
+through a builtin.
+The [Arrays in `Data`](#arrays-in-data) section considers such a case.
+
+### Slices
+
+Many array-like data structures support cheap _slices_, e.g. a function with the following
+type signature: `slice :: Integer -> Integer -> Array a -> Array a`, which produces a _view_
+of the original array between the two indices (similarly, the same can be achieved using an
+`indexArray` and a `lengthOfArray`).
+
+We do not propose to add `slice` to our builtin set, since it is very easy to build a data
+structure that supports slices on top of `array`, simply by tracking some additional
+integers to track the subset of the array that is in view.
+
+### Arrays in `Data`
+
+In [CPS-0013][1] the idea of representing the arguments to the `Data` constructor
+`Constr` as an `Array` in Plutus Core was presented as being more appropriate than
+the current list representation.
+
+A significant requirement in implementing this modification is maintaining backwards
+compatibility. Therefore, we cannot simply modify the internal representation of `Data`.
+
+One idea would be to add a new builtin such as the following:
+`unConstrDataArray :: Data -> (Integer, Array Data)`. However, this builtin will
+inevitably have linear time complexity since it is based on a list traversal, so
+it does not actually solve the original problem.
```

Comment:
It does very much solve the original problem! Because the problem is not the algorithmic complexity, it's the constants involved. `Constr` represents the constructor of a data type and chances are, it won't have many arguments, in almost all cases there's below 25 arguments, with the vast majority being below 6 (see [this](https://input-output-rnd.slack.com/archives/G01EE4NQ9U0/p1699520179968649?thread_ts=1699513075.953529&cid=G01EE4NQ9U0) slack discussion where Kenneth analyzed the hell out of it), so in practice it _is_ constant-time. But when handled on the Plutus side the constants are just really bad while as a builtin they would be fairly insignificant.

So we should definitely benchmark `unConstrDataArray`, maybe it'll perform perfectly well. Although maybe we should use the uncurried form: 

```
unConstrDataArray :: forall r. (Integer -> Array Data -> r) -> Data -> r
```

to save a builtin call (or two). Not entirely sure, needs benchmarking as well.

Reply:
@effectfully The slack link is asking me to sign in with an IOG account (which I don't have). Can you give a public summary of the discussion?

Reply:
Yes, I think this is plausible. `unConstrDataArray` may well have to be linear in the length of the array, but it might be able to have a pretty gentle slope in comparison to the linear cost of processing a builtin list.

Reply:
@fallen-icarus 

> Here's what the constructor tags look like in the redeemers for ~2,600,000 recent transactions (tag on left, frequency on right): 24 seems to be unusually popular for some reason.  I'll add the info for datums (it stlll hurts to type that) and data constants inside scripts later on
>
> ```
>   0 3362344
>   1 356404
>   2 63254
>   3 15324
>   4 27958
>   5 2138
>   6 1650
>   7 123
>   9 27
>  20 851
>  21 4329
>  22 11349
>  23 198
>  24 104886
>  30 42
>  31 58
> ```
>
> Here's the distribution for datums.  No 24s, but some 123s.  Weird.
>
> ```
>   0 10133776
>   1 1021545
>   2 73704
>   3 37826
>   4 46090
>   5 3192
>   6 2381
>   7 123
>   9 1
>  22 88
>  23 52
> 123 40
> ```
>  
> Constructor tag frequencies in data constants in recent validator scripts.
>
> ```
>   0 1321627
>   1 1944159
>   2 2173
>   6 328
> ```
>
> No, I haven't forgotten anything: that's it.

Reply:
> It does very much solve the original problem! Because the problem is not the algorithmic complexity, it's the constants involved. 

I've spent some time thinking about this claim and I think it's not accurate. We do very much detect the performance difference between `chooseList` and `chooseData` and that is reflected in the costs of the builtins despite the only substantial difference between them is `chooseList` doing an extra pattern match and maybe a few other fairly insignificant things.

So either I don't understand why `chooseData` is faster than `chooseList` (which is very possible) or indeed we can detect small changes in performance of the Haskell denotation of a Plutus builtin.

But in any case, the original statement should still be true, the overhead from multiple builtin calls is going to vastly exceed whatever inefficiencies we can have on the Haskell side.

Reply:
as a historical note, the 123's probably come from bad CBOR encoders / misunderstandings; in CBOR, "constructor 0" is encoded as 121; "constructor 1" is encoded as 122, and "constructor 2" is encoded as 123, etc.

so, someone was likely reading tools like https://cbor.nemo157.com/ and saw the 123, and then passed `123` when they should have passed `2`.

Reply:
@Quantumplation thank you very much for the explanation.

Reply:
@rphair please resolve this comment.